src/Type/Php/GetClassDynamicReturnTypeExtension.php

```diff
@@ -66,30 +67,35 @@ static function (Type $type, callable $traverse): Type {
 					return new GenericClassStringType(new ObjectType($type->getClassName()));
 				}
 
-				$objectClassNames = $type->getObjectClassNames();
-				if ($type instanceof TemplateType && $objectClassNames === []) {
-					if ($type instanceof ObjectWithoutClassType) {
-						return new GenericClassStringType($type);
-					}
-
-					return new UnionType([
-						new GenericClassStringType($type),
-						new ConstantBooleanType(false),
-					]);
-				} elseif ($type instanceof MixedType) {
-					return new UnionType([
-						new ClassStringType(),
-						new ConstantBooleanType(false),
-					]);
-				} elseif ($type instanceof StaticType) {
-					return new GenericClassStringType($type->getStaticObjectType());
-				} elseif ($objectClassNames !== []) {
-					return new GenericClassStringType($type);
-				} elseif ($type instanceof ObjectWithoutClassType) {
+				if ($type instanceof ObjectShapeType) {
 					return new ClassStringType();
 				}
 
-				return new ConstantBooleanType(false);
+				$isObject = $type->isObject();
+				if ($isObject->no()) {
```

Comment:
Answering https://github.com/phpstan/phpstan-src/pull/2350#discussion_r1443764419

It can be maybe for at least template type or mixed type